CI: Separate gofmt check from linting step

yasminvalim opened the pull request:
This PR separates the `gofmt` check from the main `golangci-lint `step.

This is necessary because the `golangci-lint-action` is hardcoded to use the `run` command and cannot be instructed to use `fmt`, making it unsuitable for formatting-only checks after the `golangci-lint `v2 update.

The new step provides clearer feedback by `gofmt -d` in the CI logs. The main linter is updated to exclude `gofmt` to prevent redundant work.

This is also related to the migration from golangci-lint `v1` to `v2` where the `gofmt `is used changed. You can check more about it here: https://golangci-lint.run/product/migration-guide/

prestist commented:
Hmm, can I get some context? is there a issue linked to this? and same question as travier?

yasminvalim commented:
So, basically, I started doing this because  we had failures in a Dependabot PR  trying to bump golangci-lint version. https://github.com/coreos/butane/actions/runs/14838167614/job/41654235381?pr=588

It happened both in Butane and Ignition, so to fix that across all Go repositories, I updated the golangci-lint version in [repo-templates](https://github.com/coreos/repo-templates/pull/292).

However, after syncing the updated template, we hit this error:
`Error: can't load config: gofmt is a formatter
`(see [CI failure](https://github.com/coreos/ignition/actions/runs/14848476477/job/41687415880?pr=2071)).

And this one:
`
Failed executing command with error: can't load config: unsupported version of the configuration: "" See https://golangci-lint.run/product/migration-guide for migration instructions. `
(see [CI failure](https://github.com/coreos/butane/actions/runs/14848480927/job/41687429009?pr=601))

According to the [migration guide](https://golangci-lint.run/product/migration-guide), gofmt is no longer treated as a linter in v2+ and should not be used with -E=gofmt. It’s now considered a formatter, and using it in args will cause errors.

To resolve this, I removed `-E=gofmt` and added gofmt as a separate step to isolate formatting from linting. This way, if cmd format fails, it fails independently. I'm open to other approaches though, looks like using gofmt -l could also work. So, do you think it makes sense? Do you have any ideas? 

I was wondering how could I test it, so I think I will add a new commit in the butane's dependabot PR to see it will works properly. I will send here my tests results ASAP.

c @travier @prestist

prestist commented:
>I was wondering how could I test it, so I think I will add a new commit in the butane's dependabot PR to see it will works properly. I will send here my tests results ASAP.

That makes sense; yes please do link it once you have it. 


>So, basically, I started doing this because we had failures in a Dependabot PR trying to bump golangci-lint version. https://github.com/coreos/butane/actions/runs/14838167614/job/41654235381?pr=588

ooof, I wonder if there is a way to ensure the CI is actually valid before merging is allowed to this repo. I wonder if we could kinda smoke test them before propagation.

ravanelli reviewed (commented):
Please add more context in the commit message and squash the commits into one

ravanelli commented on go/tests.yml:
```diff
@@ -80,7 +80,7 @@ jobs:
       with:
         version: {{ golangci_lint_version }}
-        args: -E=gofmt --timeout=30m0s
+        args: --timeout=30m0s
```
As the error, `-E is used to enable linters` and gofmt is not a linter. We are removing the format here, we should not dot it. Instead let's understand the  right replace for it.

Seems we have here:
`golangci-lint run -E=goimports` or either:
`golangci-lint run -E=gofumpt`

I'm not too familiar with the it, but let's try to understand it looking at the docs:

Seems to use goimports we will need to define the rules,  https://golangci.github.io/legacy-v1-doc/usage/linters/#goimports

I think we can try to keep the behavior for https://golangci.github.io/legacy-v1-doc/usage/linters/#gofmt
changing the cmd to support   `gofmt -l .`

ravanelli replied:
ahh, you said it here:
>> To resolve this, I removed -E=gofmt and added gofmt as a separate step to isolate formatting from linting. This way, if cmd format fails, it fails independently. I'm open to other approaches though, looks like using gofmt -l could also work. So, do you think it makes sense? Do you have any ideas?

ravanelli commented:
> Member

I think we can run it locally as a test.

ravanelli commented:
> > I was wondering how could I test it, so I think I will add a new commit in the butane's dependabot PR to see it will works properly. I will send here my tests results ASAP.
> 
> That makes sense; yes please do link it once you have it.
> 
> > So, basically, I started doing this because we had failures in a Dependabot PR trying to bump golangci-lint version. https://github.com/coreos/butane/actions/runs/14838167614/job/41654235381?pr=588
> 
> ooof, I wonder if there is a way to ensure the CI is actually valid before merging is allowed to this repo. I wonder if we could kinda smoke test them before propagation.

At least the `golangci-lint run` we can run it locally

ravanelli commented on go/tests.yml:
```diff
@@ -80,7 +80,7 @@ jobs:
 {%- endif %}
       with:
```
```suggestion
run: |
    if [ "$(gofmt -l . | wc -l)" -gt 0 ]; then
      echo "Formatting issues found:"
      gofmt -d .
      exit 1
    fi
  timeout-minutes: 30
 ```
 Maybe something around this.

yasminvalim replied:
Thanks for reviewing, Renata! I will use this code in the dependabot PR to test it.

yasminvalim commented:
Testing this changes here: https://github.com/coreos/butane/pull/588

ravanelli commented on go/tests.yaml:
```diff
@@ -1,6 +1,7 @@
 vars:
   do_go_lint: true
   do_go_mod: true
+  do_go_format: true
```
It seems to enable format for everything, what changes the current behavior that is isolated in a `if `case.

yasminvalim commented:
> Can you link the PR which you tested these changes with?
> 
> Ah found it > [coreos/butane@1a305f5](https://github.com/coreos/butane/commit/1a305f5a8e2b041b55ecae4af253bf86da4c1f5d)

This PR is kinda messy, its doing a lot of stuff like fixing lint and formatting issues, but after I merge this one the idea is to sync with this changes.
